multispecies_whale_detection/dataset.py

```diff
@@ -70,3 +92,257 @@ def new(tfrecord_filepattern: str):
+    extracted_labels = extracted_labels.write(index, label)
+    begin += hop_seconds
+    index += 1
+  return extracted_windows.stack(), extracted_labels.stack()
+
+
+def new_window_dataset(
+    tfrecord_filepattern: str,
+    context_duration_seconds: float,
+    class_names: Sequence[str],
+    sample_size: Optional[int] = None,
+    hop_seconds: Optional[float] = None,
+    min_overlap=DEFAULT_MIN_OVERLAP,
+) -> tf.data.Dataset:
+  """Parallel tf.data pipeline to extract windows from TFRecords.
+
+  This is intended for initializing a training set when sample_size is not None
+  and a validation or test set when hop_seconds is not None. Exactly one of the
+  two must be None.
+
+  Args:
+    tfrecord_filepattern: Glob pattern matching TFRecord files of serialized
+      tf.train.Examples as produced by examplegen. These represent mono audio
+      clips with sparse labels given by corresponding float endpoints and string
+      labels. batch_size
+    context_duration_seconds: The common duration of the waveforms this Dataset
+      yields.
+    class_names: String values expected in ANNOTATION_LABEL features, the order
+      of which determined class indices in the labels this dataset yields.
+    sample_size: When set, specifies that this number of context windows with
+      random starts should be extracted from each input audio. Mutually
+      exclusive with hop_seconds.
+    hop_seconds: When set, specifies that all context windows sliding over the
+      input audio at this period - that remain in bounds - should be extracted.
+      Mutually exclusive with sample_size.
+    min_overlap: Lower bound on the duration of overlap between an ANNOTATION
+      feature interval and a context window for that context window to count as
+      positive for the ANNOTATION_LABEL.
+
+  Returns:
+    Dataset of individual (context window waveform, label) tuples. All context
+    windows have the same fixed length, determined by context_window_seconds.
+    The labels are 0.0/1.0 indicators of whether an annotated interval for the
+    corresponding class overlaps the context window.
+
+  Raises:
+    ValueError: If both or neither of sample_size and hop_seconds are set.
+  """
+  if (sample_size is None) == (hop_seconds is None):
+    raise ValueError('exactly one of sample_size and hop_seconds should be set')
+  window_fn_args = dict(
+      context_duration_seconds=context_duration_seconds,
+      class_names=class_names,
+      min_overlap=min_overlap,
+  )
+  if sample_size is not None:
+    window_fn_args.update(sample_size=sample_size)
+    window_fn = functools.partial(random_windows, **window_fn_args)
+  else:
+    window_fn_args.update(hop_seconds=hop_seconds)
+    window_fn = functools.partial(sliding_windows, **window_fn_args)
+
+  def shard_fn(filename):
+    dataset = tf.data.TFRecordDataset(filename)
+    dataset = dataset.map(parse_fn)
+    dataset = dataset.map(window_fn)
+    dataset = dataset.unbatch()
```

Comment:
You may want a prefetch here to allow buffering the disk IO and parsing (similar to using readahead).

Reply:
(To avoid doubt, what follows is my fastidiousness showing.)

I might want it, but I'm not convinced it's necessary (in the mathematical sense of that word).

An expectation I maybe should have made explicit in the docstring is that a caller who is using this for training input will .shuffle().batch().prefetch(), where probably the shuffle is somewhere in (batch size, 2 * batch size) and the prefetch is 1 batch. In many cases I have found that enough to get model-bound.

TFRecordDataset(buffer_size=) is more targeted at decoupling from file reads ("disk") than is prefetch. I like defaults until they prove insufficient.

It's probably OK for record slicing, proto parsing, and Tensor filling to block on each other - because they aren't blocking any of the other parallel calls in the interleave. One big enough buffer (shuffle in train or prefetch in eval) from the caller should be enough to even it all out.

(Appendix: Is .batch().map(parse_example).unbatch() lots more efficient that .map(parse_single_example)? I don't know, but presumably there's a reason for batch and single versions of the same thing to exist. It doesn't seem intuitive that proto parsing would parallelize well.)